AzureAD/microsoft-authentication-library-for-python
# Managed Identity for Azure VM, App Service, Service Fabric, etc.

Note: This is a proof-of-concept, which means there is no guarantee that this behavior will be eventually included into MSAL Python.

There are two new APIs added.

* ~The high level API works for your confidential client which federated with a managed identity.~ This will be moved into a separated PR for its own consideration.
  ```python
   import msal
   cca = msal.ConfidentialClientApplication(
       "my_client_id",
       client_credential=msal.SystemAssignedManagedIdentity(),  # Or it can be an msal.UserAssignedManagedIdentity(client_id="guid")
       ...)
   result = cca.acquire_token_for_client(scopes["scope1", "scope2"])  # It uses scopes
  ```

* The low level API acquires token for managed identity

   ```python
   import msal, requests
   mi = msal.ManagedIdentityClient(
       msal.SystemAssignedManagedIdentity(),  # Or it can be an msal.UserAssignedManagedIdentity(client_id="guid")
       http_client=requests.Session(),  # This is a required parameter
       token_cache=msal.TokenCache(),  # Optional. In your production code, you shall persist a SerializableTokenCache https://msal-python.readthedocs.io/en/latest/#msal.SerializableTokenCache
   )
   result = mi.acquire_token_for_client(resource="resource_abc")  # It uses resource
   ```

More details of the new APIs are available [here](https://msal-python.readthedocs.io/en/mi/#managed-identity).

* In order to test this PR on Azure VM, you would need to:

  1. Create and then ssh into your Azure VM
  1. Install this proof-of-concept by `pip install --force-reinstall "git+https://github.com/AzureAD/microsoft-authentication-library-for-python.git@mi"`
  1. Write your script using the calling pattern above.

* To test this on App Service

  1. [Create your App Service with Python runtime](https://learn.microsoft.com/en-us/azure/app-service/quickstart-python?tabs=flask%2Cwindows%2Cazure-cli%2Clocal-git-deploy%2Cdeploy-instructions-azportal%2Cterminal-bash%2Cdeploy-instructions-zip-azcli#2---create-a-web-app-in-azure)
  2. [SSH into your App Service](https://learn.microsoft.com/en-us/azure/app-service/configure-linux-open-ssh-session)
  3. Follow the last two steps of Azure VM test method

* To test this on Azure Functions

  1. We have not yet tested it end-to-end, but the Managed Identity in Azure Function is expected to be the same as App Service.

* To test this on Azure Automation (we have not tested this)

  1. [Create your Automation account](https://learn.microsoft.com/en-us/azure/automation/automation-create-standalone-account?tabs=azureportal)
  3. You will need to install `msal` package. But it seems Azure Automation only supports [installing a package with its dependencies from PyPI](https://learn.microsoft.com/en-us/azure/automation/python-3-packages?tabs=py3#import-a-package-with-dependencies). This PR is not currently available from PyPI, so, we are unable to test this.
  4. After step 2, you can [create a new Python runbook](https://learn.microsoft.com/en-us/azure/automation/learn/automation-tutorial-runbook-textual-python-3) and test the Managed Identity

* To test this on Service Fabric

  1. We have not yet tested this end-to-end, but you can reference to the
     [test steps in Azure SDK](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/tests/managed-identity-live/service-fabric/service_fabric.md)

Note:

* At the end of [this internal document](https://microsoft.sharepoint.com/:w:/t/aad/devex/EXyuh-XFvj5EuoV6w8UuniUBQNg2EzQR5Ey7d_vbKPGayw?e=tTjoNQ), there are brief descriptions for the 6 variations of MIs.
* Cloud Shell's IMDS is NOT part of this PR, because we already provide a higher level API for it in `https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/420`.

Once merged, this PR will resolve #548. Also, it will officially close #487 as the callback is no longer needed.
